pkg/util/validation/limits_test.go

```diff
@@ -586,6 +587,193 @@ testuser:
 	}
 }
 
+func TestRulerMaxRulesPerRuleGroupLimits(t *testing.T) {
+	tc := []struct {
+		name      string
+		inputYAML string
+		expected  struct {
+			limit     int
+			namespace string
+		}
+	}{
+		{
+			name: "no namespace specific limit",
+			inputYAML: `
+ruler_max_rules_per_rule_group: 100
+`,
+			expected: struct {
+				limit     int
+				namespace string
+			}{limit: 100, namespace: "mynamespace"},
+		},
+		{
+			name: "zero limit for the right namespace",
+			inputYAML: `
+ruler_max_rules_per_rule_group: 100
+
+ruler_max_rules_per_rule_group_by_namespace:
+  mynamespace: 0
+`,
+			expected: struct {
+				limit     int
+				namespace string
+			}{limit: 0, namespace: "mynamespace"},
+		},
+		{
+			name: "other namespaces are not affected",
+			inputYAML: `
+ruler_max_rules_per_rule_group: 100
+
+ruler_max_rules_per_rule_group_by_namespace:
+  mynamespace: 10
+`,
+			expected: struct {
+				limit     int
+				namespace string
+			}{limit: 100, namespace: "othernamespace"},
+		},
+	}
+
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			limitsYAML := Limits{}
+			require.NoError(t, yaml.Unmarshal([]byte(tt.inputYAML), &limitsYAML))
+
+			ov, err := NewOverrides(limitsYAML, nil)
+			require.NoError(t, err)
+
+			require.Equal(t, tt.expected.limit, ov.RulerMaxRulesPerRuleGroup("user", tt.expected.namespace))
+		})
+	}
+}
+
+func TestRulerMaxRulesPerRuleGroupLimitsOverrides(t *testing.T) {
+	baseYaml := `
+ruler_max_rules_per_rule_group: 5
+
+ruler_max_rules_per_rule_group_by_namespace:
+  mynamespace: 10
+`
+
+	overrideGenericLimitsOnly := `
+testuser:
+  ruler_max_rules_per_rule_group: 333
+`
+
+	overrideNamespaceLimits := `
+testuser:
+  ruler_max_rules_per_rule_group_by_namespace:
+    mynamespace: 7777
+`
+
+	overrideGenericLimitsAndNamespaceLimits := `
+testuser:
+  ruler_max_rules_per_rule_group: 333
+
+  ruler_max_rules_per_rule_group_by_namespace:
+    mynamespace: 7777
+`
+
+	differentUserOverride := `
+differentuser:
+  ruler_max_rules_per_rule_group_by_namespace:
+    mynamespace: 500
+`
+
+	tc := []struct {
+		name           string
+		overrides      string
+		inputNamespace string
+		expectedLimit  int
+	}{
+		{
```

Comment:
super-nit: if order is not important, `map[string]testCase` is preferrable, to avoid duplicate names.

Reply:
Done in https://github.com/grafana/mimir/pull/8378/commits/426289646b3ba7479d456e82f20b74c3432a9bd2 - I actually changed all of the tests that followed the above pattern.